File: src/main/java/uk/co/unclealex/music/musicbrainz/MusicBrainzClientImpl.java
/**
 * Copyright 2012 Alex Jones
 *
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 * 
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.    
 *
 * @author unclealex72
 *
 */

package uk.co.unclealex.music.musicbrainz;

import java.util.List;

import javax.inject.Inject;

import org.apache.commons.lang3.tuple.ImmutablePair;
import org.jdom2.Document;
import org.jdom2.Element;
import org.jdom2.Namespace;
import org.jdom2.filter.ElementFilter;

import uk.co.unclealex.music.configuration.User;

import com.google.common.base.Joiner;
import com.google.common.base.Predicate;
import com.google.common.collect.Iterables;
import com.google.common.collect.Lists;
import com.sun.jersey.api.client.WebResource;

/**
 * The Class MusicBrainzClientImpl.
 *
 * @author alex
 */
public class MusicBrainzClientImpl implements MusicBrainzClient {

  /**
   * The maximum number of release IDs allowed in a PUT or DELETE request.
   */
  private static final int RELEASE_PATH_LIMIT = 400;

  /** 
   * The maximum size of a query.
   */
  private static final int LIMIT = 100;

  /** 
   * The name of the user's collection. 
   */
  private static final String ALL_MY_MUSIC = "All my music";

  /**
   * The namespace for XML documents returned by MusicBrainz.
   */
  private static Namespace MUSICBRAINZ_NS = Namespace.getNamespace("http://musicbrainz.org/ns/mmd-2.0#");

  /**
   * The {@link MusicBrainzWebResourceFactory} used to talk to MusicBrainz with
   * Jersey.
   */
  private final MusicBrainzWebResourceFactory musicBrainzWebResourceFactory;

  /**
   * Instantiates a new music brainz client impl.
   *
   * @param musicBrainzWebResourceFactory the music brainz web resource factory
   */
  @Inject
  public MusicBrainzClientImpl(MusicBrainzWebResourceFactory musicBrainzWebResourceFactory) {
    super();
    this.musicBrainzWebResourceFactory = musicBrainzWebResourceFactory;
  }

  /**
   * {@inheritDoc}
   * 
   */
  @Override
  public List<String> getRelasesForOwner(User user) throws NoCollectionException {
    ImmutablePair<String, Integer> collectionInfo = findUsersCollection(user);
    String collectionId = collectionInfo.getLeft();
    int collectionSize = collectionInfo.getRight();
    List<String> releases = Lists.newArrayListWithCapacity(collectionSize);
    for (int offset = 0; offset < collectionSize; offset += LIMIT) {
      Document document = releases(user, collectionId)
          .queryParam("offset", Integer.toString(offset))
          .queryParam("limit", Integer.toString(LIMIT)).get(Document.class);
      for (Element el : document.getDescendants(new ElementFilter("release", MUSICBRAINZ_NS))) {
        String releaseId = el.getAttributeValue("id");
        releases.add(releaseId);
      }
    }
    return releases;
  }

  /**
   * Get the web resource for a users collection.
   * @param user The user who is making the request.
   * @param collectionId The id of the collection to retrieve or alter.
   * @return A web resource for a user's collection.
   */
  protected WebResource releases(User user, String collectionId) {
    return getMusicBrainzWebResourceFactory()
        .webResource(user)
        .path("collection")
        .path(collectionId)
        .path("releases");
  }

  /**
   * {@inheritDoc}
   */
  @Override
  public void addReleases(User user, Iterable<String> newReleaseIds) throws NoCollectionException {
    WebResourceAction put = new WebResourceAction() {
      @Override
      public void act(WebResource webResource) {
        webResource.put();
      }
    };
    alterReleases(user, newReleaseIds, put);
  }

  /**
   * {@inheritDoc}
   */
  @Override
  public void removeReleases(User user, Iterable<String> oldReleaseIds) throws NoCollectionException {
    WebResourceAction delete = new WebResourceAction() {
      @Override
      public void act(WebResource webResource) {
        webResource.put();
      }
    };
    alterReleases(user, oldReleaseIds, delete);
  }
  
  /**
   * Alter a user's collection of releases.
   * @param user The user whose collection is being searched.
   * @param releaseIds The ids of the affected releases.
   * @param action The web method to execute.
   * @throws NoCollectionException 
   */
  private void alterReleases(User user, Iterable<String> releaseIds, WebResourceAction action) throws NoCollectionException {
    String collectionId = findUsersCollection(user).getLeft();
    for (Iterable<String> partition : Iterables.partition(releaseIds, RELEASE_PATH_LIMIT)) {
      String releasesPath = Joiner.on(';').join(partition);
      action.act(releases(user, collectionId).path(releasesPath));
    }
  }

  /**
   * An interface to encapsulate a method being called on a web resource.
   * @author alex
   *
   */
  interface WebResourceAction {
    
    public void act(WebResource webResource);
  }
  /**
   * Find a user's unique collection.
   * 
   * @param user
   *          The user whose collection is being searched for.
   * @return An {@link ImmutablePair} containing the ID of the collection and
   *         the number of entries it requires.
   * @throws NoCollectionException
   *           thrown if a unique collection cannot be found.
   */
  public ImmutablePair<String, Integer> findUsersCollection(User user) throws NoCollectionException {
    Document doc = getMusicBrainzWebResourceFactory().webResource(user).path("collection").get(Document.class);
    List<Element> collectionElements =
        Lists.newArrayList((Iterable<Element>) doc.getDescendants(new ElementFilter("collection", MUSICBRAINZ_NS)));
    int size = collectionElements.size();
    Element collectionElement;
    if (size == 0) {
      throw new NoCollectionException("No collections found.");
    }
    else if (size == 1) {
      collectionElement = collectionElements.get(0);
    }
    else {
      Predicate<Element> isAllMyMusicPredicate = new Predicate<Element>() {
        public boolean apply(Element element) {
          Element nameElement = element.getChild("name", MUSICBRAINZ_NS);
          return nameElement != null && ALL_MY_MUSIC.equals(nameElement.getTextNormalize());
        }
      };
      collectionElement = Iterables.find(collectionElements, isAllMyMusicPredicate, null);
      if (collectionElement == null) {
        throw new NoCollectionException("Cannot find a unique collection.");
      }
    }
    String releaseId = collectionElement.getAttributeValue("id");
    int releaseCount =
        Integer.parseInt(collectionElement.getChild("release-list", MUSICBRAINZ_NS).getAttributeValue("count"));
    return new ImmutablePair<String, Integer>(releaseId, releaseCount);
  }

  /**
   * Gets the {@link MusicBrainzWebResourceFactory} used to talk to MusicBrainz with Jersey.
   *
   * @return the {@link MusicBrainzWebResourceFactory} used to talk to MusicBrainz with Jersey
   */
  public MusicBrainzWebResourceFactory getMusicBrainzWebResourceFactory() {
    return musicBrainzWebResourceFactory;
  }

}

File: app/common/musicbrainz/MusicBrainzClientImpl.scala
/*
 * Copyright 2014 Alex Jones
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *       http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package common.musicbrainz

import java.io.InputStream
import java.lang.annotation.Annotation
import java.lang.reflect.Type
import javax.ws.rs.core.{HttpHeaders, MediaType, MultivaluedMap}
import javax.ws.rs.ext.MessageBodyReader

import com.sun.jersey.api.client.filter.ClientFilter
import com.sun.jersey.api.client.{ClientRequest, ClientResponse, WebResource}
import com.sun.jersey.client.apache.ApacheHttpClient
import com.sun.jersey.client.apache.config.ApacheHttpClientConfig._
import com.sun.jersey.client.apache.config.DefaultApacheHttpClientConfig
import com.typesafe.scalalogging.StrictLogging
import common.configuration.{PlayConfiguration, User}
import play.api.Configuration

import scala.xml.{Elem, XML}

/**
 * The Class MusicBrainzClientImpl.
 *
 * @author alex
 */
class MusicBrainzClientImpl(val musicBrainzHost: String) extends MusicBrainzClient with StrictLogging {

  case class Collection(id: String, name: String, size: Int)

  override def relasesForOwner(user: User): Traversable[String] = {
    implicit val implicitUser = user
    loadCollection(defaultCollection)
  }

  def defaultCollection(implicit user: User): Collection = {
    def collectionsResponse = MusicBrainzWebResource(user).get(classOf[Elem])
    findDefaultCollection(user)(collectionsResponse)
  }

  def findDefaultCollection(implicit user: User): Elem => Collection = (xml: Elem) => {
    logger.debug(s"Looking for default connection\n${xml}")
    val collections = for {
      collection <- xml \\ "collection"
      releaseList <- collection \ "release-list"
    } yield
      Collection(collection \@ "id", (collection \ "name").text, (releaseList \@ "count").toInt)
    collections.find(_.name == ALL_MY_MUSIC) getOrElse {
      throw new IllegalStateException(s"User ${user.name} does not have a collection called $ALL_MY_MUSIC")
    }
  }

  def loadCollection(collection: Collection)(implicit user: User): Traversable[String] = {
    val parts = 0 until collection.size by LIMIT map { offset => loadCollection(collection, offset)}
    parts.flatten
  }

  def loadCollection(collection: Collection, offset: Int)(implicit user: User): Traversable[String] = {
    val webResource = MusicBrainzWebResource(user)
    val collectionResponse =
      webResource.path(collection.id).path("releases").queryParam("offset", offset.toString).queryParam("limit", LIMIT.toString).get(classOf[Elem])
    collectionResponse \\ "release" map (_ \@ "id")
  }

  /**
   * Add releases to an owner's collection.
   * @param user The user whose collection needs changing.
   * @param newReleaseIds The new releases to add to the user's collection.
   * @throws thrown if a unique collection cannot be found.
   */
  override def addReleases(user: User, newReleaseIds: Set[String]): Unit = {
    alterReleases(wr => cl => wr.put(cl), newReleaseIds)(user)
  }

  /**
   * Remove releases from an owner's collection.
   * @param user The user whose collection needs changing.
   * @param oldReleaseIds The old releases to remove from the user's collection.
   * @throws thrown if a unique collection cannot be found.
   */
  override def removeReleases(user: User, oldReleaseIds: Set[String]): Unit = {
    alterReleases(wr => cl => wr.delete(cl), oldReleaseIds)(user)
  }

  def alterReleases(method: WebResource => Class[String] => String, releaseIds: Set[String])(implicit user: User): Unit = {
    val alterRelease: String => Unit = { collectionId =>
      val groups = releaseIds.toList.grouped(RELEASE_PATH_LIMIT)
      groups.foreach { releaseIds =>
        val webResource = MusicBrainzWebResource(user).path(collectionId).path("releases").path(releaseIds.mkString(";"))
        method(webResource)(classOf[String])
      }
    }
    alterRelease(defaultCollection.id)
  }

  object MusicBrainzWebResource extends MessageBodyReader[Elem] {

    override def isReadable(`type`: Class[_], genericType: Type, annotations: Array[Annotation], mediaType: MediaType): Boolean = {
      classOf[Elem] == `type`
    }

    override def readFrom(
                           `type`: Class[Elem],
                           genericType: Type,
                           annotations: Array[Annotation],
                           mediaType: MediaType,
                           httpHeaders: MultivaluedMap[String, String],
                           entityStream: InputStream): Elem = XML.load(entityStream)

    def apply(user: User): WebResource = {
      val cc = new DefaultApacheHttpClientConfig()
      cc.getSingletons().add(this)
      cc.getProperties().put(PROPERTY_PREEMPTIVE_AUTHENTICATION, java.lang.Boolean.TRUE)
      cc.getState().setCredentials("musicbrainz.org", null, -1, user.musicBrainzUserName, user.musicBrainzPassword)
      val client = ApacheHttpClient.create(cc)
      client.addFilter(new MusicBrainzRetryFilter(100, 5))
      val userAgentClientFilter = new ClientFilter() {
        override def handle(cr: ClientRequest): ClientResponse = {
          cr.getHeaders().add(HttpHeaders.USER_AGENT, USER_AGENT)
          getNext().handle(cr)
        }
      }
      client.addFilter(userAgentClientFilter)
      client.resource(s"http://${musicBrainzHost}/ws/2/collection")
    }

  }

}

class PlayConfigurationMusicBrainzClient(override val configuration: Configuration) extends PlayConfiguration[MusicBrainzClient](configuration) with MusicBrainzClient {

  override def load(configuration: Configuration): Option[MusicBrainzClient] =
    configuration.getString("musicbrainzHost").map(host => new MusicBrainzClientImpl(host))

  override def addReleases(user: User, newReleaseIds: Set[String]): Unit = result.addReleases(user, newReleaseIds)

  override def removeReleases(user: User, oldReleaseIds: Set[String]): Unit = result.removeReleases(user, oldReleaseIds)

  override def relasesForOwner(user: User): Traversable[String] = result.relasesForOwner(user)
}